### PR TITLE
cmd/evm: make evm t8n handle post-merge transitions

### LINKED ISSUE #24003: t8ntool support merge rules
Add new fork to t8ntool `--fork Merge` for example.
That tells the evm to work with PoS mode. meaning env section follows the nonce, mixhash, difficulty requrements. 
random bytecode and so on.

### PR DESCRIPTION
**Pull request by holiman**
This PR closes #24003 . 
This adds the ability to run `--state.fork=Merged`, and have post-merge rules apply. When doing so, it also requires the input `env` to contain `currentRandom`, and enforces the `currentDifficulty` to be omitted or zero. 

This also adds a method to the chainconfig: `IsMerged(num *big.Int)`. It's meant to be used only in some scenarios, like when `t8n` is testing transitions, or when we execute state tests post-merge. I don't think the method should be used by production code, which must be more careful with ttd checks and the origin of the block (engine api or network). I'm not sure if having this method is "something we need anyway" or something we should avoid?

### CODE REVIEW

**Comment by MariusVanDerWijden** [comment]
Hmm one small thing: the mergeFork might occur before or after the actual merge (it's only intended for the p2p forkid)

**Comment by holiman** [comment]
>  the mergeFork might occur before or after the actual merge (it's only intended for the p2p forkid)

Oh? I thought it was something we would set after we know the actual merge block number (combined with a hash). 

But yeah, I know that it's not correct to treat TheMerge like any other blocknumber-based-fork. But for testing post-merge, we want to be able to say "we're post-merge now trust me" -- rather than just set some fields here and there and hope that the execution engine infers the correct ruleset. 

But I'm not sure what the best approach for that is.

**Comment by winsvega** [comment]
here it is only about running evm in merge rules. (with currentRandom and other accepted EIPs) 
all the rest is blockchain logic we handle with the tests when generating the blocks. like transition tests.

File: cmd/evm/internal/t8ntool/transition.go [hunk 1]
```diff
@@ -252,11 +252,20 @@ func Transition(ctx *cli.Context) error {
 			return NewError(ErrorConfig, errors.New("EIP-1559 config but missing 'currentBaseFee' in env section"))
 		}
 	}
-	// Sanity check, to not `panic` in state_transition
-	if prestate.Env.Random != nil && !chainConfig.IsLondon(big.NewInt(int64(prestate.Env.Number))) {
-		return NewError(ErrorConfig, errors.New("can only apply RANDOM on top of London chainrules"))
-	}
-	if env := prestate.Env; env.Difficulty == nil {
+
+	if env := prestate.Env; chainConfig.IsMerged(big.NewInt(int64(env.Number))) {
```
Comment (holiman) [inline cmd/evm/internal/t8ntool/transition.go:256]: This `IsMerged` makes the semantics simple, but I could skip that method and just do 
```
isMerged = (env.Random == nil) && (env.Difficulty = nil || env.Difficulty.BitLen()==0)
```
Would that be preferred?

Comment (MariusVanDerWijden) [reply cmd/evm/internal/t8ntool/transition.go:256]: Yes, I think I prefer this over making decisions based on a fork that "we should not base any changes other than the forkid" on.

**Comment by holiman** [comment]
Fixed now to not touch the core parts, just use the terminal total difficulty check

**Comment by holiman** [comment]
@MariusVanDerWijden PTAL

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM

**Pull request merged**